low_level_platform/impl/src/lf_STM32f4_support.c

```diff
@@ -0,0 +1,238 @@
+#if defined(PLATFORM_STM32F4)
+
+#include "platform/lf_STM32f4_support.h"
+#include "low_level_platform.h"
+#include "tag.h"
+
+//  + -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- +
+//  | Important defines and global variables
+//  + -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- +
+
+static volatile bool _lf_sleep_interrupted = false;
+static volatile bool _lf_async_event = false;
+
+// nested critical section counter
+static uint32_t _lf_num_nested_crit_sec = 0;
+
+// Timer upper half (for overflow)
+static uint32_t _lf_time_us_high = 0;
+
+#define LF_MIN_SLEEP_NS 10
+
+// Combine 2 32bit works to a 64 bit word (Takes from nrf52 support)
+#define COMBINE_HI_LO(hi, lo) ((((uint64_t)hi) << 32) | ((uint64_t)lo))
+
+void lf_SystemClock_Config();
+void Error_Handler();
+
+//  + -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- +
+//  | Code for timer functions
+//  + -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- -- +
+// We use timer 5 for our clock (probably better than dealing with sysTick)
+void _lf_initialize_clock(void) {
+  // Standard initializations from generated code
+  HAL_Init();
+  lf_SystemClock_Config();
+
+  // Configure TIM5 as our 32-bit clock timer
+  __HAL_RCC_TIM5_CLK_ENABLE(); // initialize counter
+  TIM5->CR1 = TIM_CR1_CEN;     // enable counter
+
+  // set prescaler to (16 - 1) = 15
+  // CPU runs a 16MHz so timer ticks at 1MHz
+  //      Which means period of 1 microsecond
+  TIM5->PSC = 15;
+
+  // Setup ISR to increment upper bits
+  TIM5->DIER |= TIM_DIER_CC1IE;
+  NVIC_EnableIRQ(TIM5_IRQn);
+
+  /* This is to make the Prescaler actually work
+   *  For some reason, the Prescaler only kicks in once the timer has reset once.
+   *  Thus, the current solution is to manually ret the value to a really large
+   *  and force it to reset once. Its really jank but its the only way I could
+   *  find to make it work
+   */
+  TIM5->CNT = 0xFFFFFFFE;
+}
+
+// ISR for handling timer overflow -> We increment the upper timer
+void TIM5_IRQHandler(void) {
+  if (TIM5->SR & (1 << 1)) {
+    TIM5->SR &= ~(1 << 1);
+    _lf_time_us_high += 1;
+  }
+}
+
+/**
+ * Write the time since boot into time variable
+ */
+int _lf_clock_gettime(instant_t* t) {
+  // Timer is cooked
+  if (!t) {
+    return 1;
+  }
+  // Get the current microseconds from TIM5
+  uint32_t _lf_time_us_low = TIM5->CNT;
```

Comment:
There is a tricky race condition here that appears during the clock cycles where a counter overflow happens. Imagine the following scenario:
You read TIM5->CNT when the value is 0xffffffff (i.e. UINT32_MAX). Then there is an overflow and `_lf_time_us_high` is incremented before you read it a few lines below here. Then you end up with the wrong value.

Two ways to solve it, either duplicate read-outs (see NRF52 code), or disable interrupts (so that the overflow IRQ is not handled)